main/src/ca/uwaterloo/flix/language/phase/unification/ContextReduction.scala

```diff
@@ -0,0 +1,115 @@
+        case Result.Err(_) => false
+      }
+    }
+  }
+
+  /**
+    * Removes the type constraints which are entailed by the others in the list.
+    */
+  private def simplify(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): List[TypedAst.TypeConstraint] = {
+
+    @tailrec
+    def loop(acc: List[TypedAst.TypeConstraint], tconstrs0: List[TypedAst.TypeConstraint]): List[TypedAst.TypeConstraint] = tconstrs0 match {
+      case Nil => acc
+      case head :: tail if entail(instances, acc ++ tail, head) => loop(acc, tail)
+      case head :: tail => loop(head :: acc, tail)
+    }
+
+    loop(Nil, tconstrs0)
+  }
+
+  /**
+    * Normalizes a list of type constraints, converting to head-normal form and removing semantic duplicates.
+    */
+  private def reduce(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstrs0: List[TypedAst.TypeConstraint])(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    for {
+      tconstrs <- Result.sequence(tconstrs0.map(toHeadNormalForm(instances, _)))
+    } yield simplify(instances, tconstrs.flatten)
+  }
+
+  /**
+    * Converts the type constraint to head-normal form, i.e. `a[X1, Xn]`, where `a` is a variable and `n >= 0`.
+    */
+  private def toHeadNormalForm(instances: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    if (isHeadNormalForm(tconstr.arg)) {
+      List(tconstr).toOk
+    } else {
+      byInst(instances, tconstr)
+    }
+  }
+
+  /**
+    * Returns the list of constraints that hold if the given constraint `tconstr` holds, using the constraints on available instances.
+    */
+  private def byInst(instances0: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+    val matchingInstances = instances0(tconstr.sym).toList
+
+    def tryInst(inst: ResolvedAst.Instance): Result[List[TypedAst.TypeConstraint], UnificationError] = {
+      for {
+        subst <- Unification.unifyTypes(tconstr.arg, inst.tpe)
+      } yield inst.tconstrs.map(subst(_))
+    }
+
+    matchingInstances.map(tryInst).find {
+      case Result.Ok(_) => true
+      case Result.Err(_) => false
+    } match {
+      case Some(Result.Ok(tconstrs)) => tconstrs.toOk
+      case _ => UnificationError.NoMatchingInstance(tconstr.sym, tconstr.arg).toErr
+    }
+  }
+
+  /**
+    * Returns the list of constraints that hold if the given constraint `tconstr` hold, using the superclasses of the constraint.
+    * (Currently only returns the given constraint since superclasses are not yet implemented.)
+    */
+  private def bySuper(_instances0: MultiMap[Symbol.ClassSym, ResolvedAst.Instance], tconstr: TypedAst.TypeConstraint)(implicit flix: Flix): List[TypedAst.TypeConstraint] = {
```

Comment:
OK, I understand.

Reply:
THIH says that we must not have cyclic inheritance. I guess that check does not become relevant until we add super classes.